metrics/collectors/shared.lua

```diff
@@ -72,6 +74,32 @@ function Shared:dec(num, label_pairs)
     self.label_pairs[key] = label_pairs
 end
 
+--- Measure latency of function call
+--
+-- @param label_pairs either table with labels or function to generate labels.
+--      If function is given its called with the results of pcall.
+-- @param fn function for pcall to instrument
+-- ... - args for function fn
+function Shared:observe_latency(label_pairs, fn, ...)
+    local start_time = clock.monotonic()
+    local ok, result, err_msg = pcall(fn, ...)
+    local latency = clock.monotonic() - start_time
+
+    xpcall(function()
+        if type(label_pairs) == 'function' then
+            label_pairs = label_pairs(ok, result, err_msg)
+        end
+        self:observe(latency, label_pairs)
+    end, function(err)
+        log.error(debug.traceback('Saving metrics failed: ' .. tostring(err)))
```

Comment:
I don't like the idea when metric writes something to logs. It's easy way to spam user logs

Reply:
And maybe `xpcall` body should be reworked... For each call you create new closure that leads to additional pressure on GC.

Reply:
Any suggestion how to get any debug info when collector fails? This log is called only if collector fails. Default collectors just add a records to couple of tables.

Regarding to closures, I think this overhead is not comparable to the latency of instrumented functions, Isn't it?

Reply:
> Any suggestion how to get any debug info when collector fails? This log is called only if collector fails. Default collectors just add a records to couple of tables.

I have similar discussion here - https://github.com/tarantool/metrics/pull/69#discussion_r432415901
IMO it's inappropriate to have an ability to spam user's logs.

That about errors... I have single thought for this - implement smth similar to "errno" variables and use it for diagnostics purposes.

> Regarding to closures, I think this overhead is not comparable to the latency of instrumented functions, Isn't it?

Maybe yes, but anyway seems if we can write code with the minimum possible overhead, then we should do so